pypiserver/_app.py

```diff
@@ -292,59 +276,62 @@ def simpleindex():
     return template(tmpl, links=links)
 
 
-@app.route("/simple/:prefix/")
+@app.route("/simple/:project/")
 @auth("list")
-def simple(prefix=""):
-    # PEP 503: require normalized prefix
-    normalized = core.normalize_pkgname_for_url(prefix)
-    if prefix != normalized:
-        return redirect("/simple/{0}/".format(normalized), 301)
-
-    files = sorted(
-        core.find_packages(packages(), prefix=prefix),
+def simple(project):
+    # PEP 503: require normalized project
+    normalized = normalize_pkgname_for_url(project)
+    if project != normalized:
+        return redirect(f"/simple/{normalized}/", 301)
+
+    packages = sorted(
+        core.find_project_packages(project),
         key=lambda x: (x.parsed_version, x.relfn),
     )
-    if not files:
+    if not packages:
         if config.redirect_to_fallback:
-            return redirect(f"{config.fallback_url.rstrip('/')}/{prefix}/")
+            return redirect(f"{config.fallback_url.rstrip('/')}/{project}/")
         return HTTPError(404, f"Not Found ({normalized} does not exist)\n\n")
 
-    fp = request.custom_fullpath
+    current_uri = request.custom_fullpath
+
     links = [
         (
-            os.path.basename(f.relfn),
-            urljoin(fp, f"../../packages/{f.fname_and_hash(config.hash_algo)}"),
+            os.path.basename(pkg.relfn),
+            urljoin(current_uri, f"../../packages/{pkg.fname_and_hash}"),
         )
-        for f in files
+        for pkg in packages
     ]
+
     tmpl = """\
     <html>
         <head>
-            <title>Links for {{prefix}}</title>
+            <title>Links for {{project}}</title>
         </head>
         <body>
-            <h1>Links for {{prefix}}</h1>
+            <h1>Links for {{project}}</h1>
             % for file, href in links:
                  <a href="{{href}}">{{file}}</a><br>
             % end
         </body>
     </html>
     """
-    return template(tmpl, prefix=prefix, links=links)
+    return template(tmpl, project=project, links=links)
 
 
 @app.route("/packages/")
 @auth("list")
 def list_packages():
     fp = request.custom_fullpath
-    files = sorted(
-        core.find_packages(packages()),
+    packages = sorted(
+        core.get_all_packages(),
         key=lambda x: (os.path.dirname(x.relfn), x.pkgname, x.parsed_version),
     )
+
     links = [
-        (f.relfn_unix, urljoin(fp, f.fname_and_hash(config.hash_algo)))
-        for f in files
+        (pkg.relfn_unix, urljoin(fp, pkg.fname_and_hash)) for pkg in packages
     ]
```

Comment:
Could we try replacing this list comprehension (`[]`) with a generator comprehension (`()`)? We definitely need to allocate a list for sorting, but here I think we may be able to be lazy and just pass a generator into the template call, which should save some RAM on systems handling a lot of packages

Reply:
Yeah sure :)

Good catch on the `fname_and_hash`. Apparently no test caught that